tests/test_geometry.py

```diff
@@ -175,21 +175,119 @@ def test_stepExpansion_fun2par(projection, func):
     # Compare projection with fun2par results
     assert np.allclose(p, qa_f.parameters)
 
-def test_KL_expansion():
+@pytest.mark.parametrize("num_modes",[1, 10, 20, 25])
+def test_KL_expansion(num_modes):
     """Check KL expansion geometry correctness"""
     N = 20
-    grid = np.linspace(0,1,N)
+    grid = np.linspace(0, 1, N)
     decay_rate = 2.5
     normalizer = 12
     geom = cuqi.geometry.KLExpansion(grid,
                                      decay_rate=decay_rate,
-				     normalizer=normalizer)
+                                     normalizer=normalizer,
+                                     num_modes=num_modes)
+    if num_modes > len(grid):
+        num_modes = len(grid)
+
     p = np.random.randn(N)
-    f_geom = geom.par2fun(p)
-    f_expected = _inverse_sin_discrete_transform_KL(p, N, decay_rate, normalizer)
+    f_geom = geom.par2fun(p[:num_modes])
 
-    assert np.allclose(f_geom, f_expected)
+    p[num_modes:] = 0
+    f_expected = _inverse_sin_discrete_transform_KL(
+        p, N, decay_rate, normalizer)
+
+    assert np.allclose(f_geom, f_expected) and\
+        len(geom.coefs) == geom.par_dim and\
+        geom.par_dim == geom.num_modes and\
+        len(geom.grid) == geom.fun_dim
```

Comment:
I think seperate asserts will make the error easier to catch in future.

Reply:
👍 done